aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -144,6 +144,14 @@ impl<V: TransactionWrite> DataRead<V> {
             (_, _) => unreachable!("{:?}, {:?} must be covered", self_kind, kind),
         })
     }
+
+    pub(crate) fn from_value_with_layout(version: Version, value: ValueWithLayout<V>) -> Self {
+        match value {
+            // If value was never exchanged, then it can be the highest one without full value.
+            ValueWithLayout::RawFromStorage(v) => DataRead::Metadata(v.as_state_value_metadata()),
```

Comment:
Can only metadata be stored in `ValueWithLayout::RawFromStorage`? In that case, why not call it `ValueWithLayout::Metadata`?

Reply:
this is unrelated to resource groups and metadata. 

regular values can be stored in RawFromStorage, if for example only resource_exists is called (which doesn't provide layout). check out the newly added tests

Reply:
this is critical piece - if we have RawFromStorage, we cannot be leaking "value" outside. So the kind needs to be the next lower - which is metadata, and that allows downcast/contains to work correctly.